nep-20.mediawiki

```diff
@@ -0,0 +1,419 @@
+    getAccounts(): Promise<Account[]>;
+    getApplicationLog(txid: UInt256): Promise<ApplicationLog>;
+    getBalance(asset: UInt160, account?: UInt160): Promise<Integer>;
+    getBlock(hash: UInt256): Promise<Block>;
+    getBlock(index: number): Promise<Block>;
+    getBlockCount(): Promise<number>;
+    getStorage(hash: UInt160, key: Base64Encoded): Promise<Base64Encoded>;
+    getTokenInfo(hash: UInt160): Promise<Token>;
+    getTransaction(txid: UInt256): Promise<Transaction>;
+    invoke(invocations: InvocationArguments[], signers?: Signer[], suggestedSystemFee?: Integer): Promise<UInt256>;
+    makeTransaction(invocations: InvocationArguments[], signers?: Signer[], suggestedSystemFee?: Integer): Promise<ContractParametersContext>;
+    relay(context: ContractParametersContext): Promise<UInt256>;
+    send(asset: UInt160, amount: Integer, to: UInt160, from?: UInt160): Promise<UInt256>;
+    sign(context: ContractParametersContext): Promise<ContractParametersContext>;
+}
+</pre>
+
+===Errors===
+
+<pre>
+export type ErrorType =
+    "UNSUPPORTED" |         // The requested feature or operation is not supported.
+    "INVALID" |             // The input data is in an invalid format.
+    "NOTFOUND" |            // The requested data doesn't exist.
+    "FAILED" |              // The contract execution failed.
+    "TIMEOUT" |             // The requested operation was cancelled due to timeout.
+    "CANCELED" |            // The requested operation was cancelled by the user.
+    "INSUFFICIENT_FUNDS" |  // The requested operation failed due to insufficient balance.
+    "RPC_ERROR" |           // An exception was thrown by the RPC server.
+    "UNKNOWN";              // An unknown error has occurred.
+
+export interface Error {
+    type: ErrorType;
+    message: string;
+    detail?: any;
+}
+
+export interface FailedError extends Error {
+    type: "FAILED";
+    detail: InvocationResult;
+}
+</pre>
+
+===How to obtain an instance of IDapiProvider interface?===
+
+In order for dapps to have a unified method of obtaining <code>IDapiProvider</code> instances, all providers must trigger the <code>Neo.DapiProvider.ready</code> event of the <code>window</code> object when they are ready.
+
+The front end of the dapp can listen to the event and obtain the <code>IDapiProvider</code> instance from the <code>detail</code> property of the event.
+
+<pre>
+window.addEventListener("Neo.DapiProvider.ready", e => {
+    var provider = e.detail.provider;
+});
+</pre>
+
+==Rationale==
+
+This protocol will allow dApp developers to create applications that interact with the NEO blockchain without having to be concerned about managing a full wallet within their application or the details related to handing transaction creation or broadcasting. This will also allow dApps to allow users to transact in a secure fashion that does not require sharing of their private key.

```

Comment:
Is this intended to be a protocol or an API? the specification above appears to be a Typescript API, but here it's described as a protocol (which would imply the possibility of multiple implementations in different programming languages)

Reply:
Currently, the front end of most dapps is Javascript running in browsers.

Reply:
> Currently, the front end of most dapps is Javascript running in browsers.

This doesn't answer my "is this intended to be a protocol or api" question

Reply:
The specification describes the interfaces, so it is a protocol. Different wallet providers can have different implementations, but the interaction protocol with dapps is the same.